### PR TITLE
Feat/GH441 grapher tool; fixes #441

### LINKED ISSUE #441: First pass implementation of Grapher plugin.

### PR DESCRIPTION
**Pull request by nathanatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/pulls) open
- [x] I have verified that I am following the CDP4-IME [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/CDP4-IME-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
fixes #441

- [x] Make it reactive
  - [x] update properties when prop are changed
  - [x] redraw everything whenever ElementUsage are removed or added

- [x] Org the tree correctly
- [x] Remove ClassName and ThingIcon
- [x] change grapher icon
- [x] Unit tests
- [x] comma space category
- [x] Store layout to re-apply on redraw
- [x] Reorder the nested element so they are ordered by alphanumerical order like in the product view
- [x] Redraw whole tree whenever element => deleted created

### CODE REVIEW

File: CDP4Composition/Converters/CategoryFromElementUsageCollectionLastElementToTextConverter.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//    GNU Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+
```
Comment (samatrhea) [inline CDP4Composition/Converters/CategoryFromElementUsageCollectionLastElementToTextConverter.cs:27]: lines

File: CDP4Composition/Converters/CategoryFromElementUsageCollectionLastElementToTextConverter.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Linq;
+    using System.Windows.Data;
+
+    using CDP4Common.EngineeringModelData;
+    using CDP4Common.SiteDirectoryData;
+
+    public class CategoryFromElementUsageCollectionLastElementToTextConverter : IValueConverter
```
Comment (samatrhea) [inline CDP4Composition/Converters/CategoryFromElementUsageCollectionLastElementToTextConverter.cs:39]: please put docs on the class to explain the purpose, somthing like:

`The purpose of the CategoryFromElementUsageCollectionLastElementToTextConverter class is to...`

File: CDP4Composition/Converters/CategoryFromElementUsageCollectionLastElementToTextConverter.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        /// </param>
+        /// <param name="culture">
+        /// The culture information.
+        /// </param>
+        /// <returns>
+        /// The <see cref="List{T}"/> of <see cref="T"/> containing the same objects as the input collection.
+        /// </returns>
+        public virtual object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
+        {
+            throw new NotImplementedException();
```
Comment (samatrhea) [inline CDP4Composition/Converters/CategoryFromElementUsageCollectionLastElementToTextConverter.cs:89]: please make this a `NotSupportedException`

File: CDP4Grapher.Tests/Behaviors/GrapherOrgChartBehaviorTestFixture.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+//    The CDP4-IME Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//    GNU Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
```
Comment (samatrhea) [inline CDP4Grapher.Tests/Behaviors/GrapherOrgChartBehaviorTestFixture.cs:26]: lines

File: CDP4Composition/Converters/CategoryFromElementUsageCollectionLastElementToTextConverter.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="CategoryFromElementUsageCollectionLastElementToTextConverter.cs" company="RHEA System S.A.">
+//    Copyright (c) 2015-2020 RHEA System S.A.
+//
+//    Author: Sam Gerené, Alex Vorobiev, Merlin Bieze, Naron Phou, Patxi Ozkoidi, Alexander van Delft,
```
Comment (samatrhea) [inline CDP4Composition/Converters/CategoryFromElementUsageCollectionLastElementToTextConverter.cs:5]: please remove Merlin, Naron and Patxi.

File: CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+                    this.ApplyMindMapLayout(d);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+
+            this.CurrentLayout = (layout, direction);
+        }
+
+        private void ItemsChanged(object sender, DiagramItemsChangedEventArgs e)
```
Comment (samatrhea) [inline CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs:135]: docs

File: CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        /// </summary>
+        protected override void OnDetaching()
+        {
+            this.AssociatedObject.DataContextChanged -= this.OnDataContextChanged;
+            this.AssociatedObject.ItemsChanged -= this.ItemsChanged;
+            this.AssociatedObject.Loaded -= this.Loaded;
+            base.OnDetaching();
+        }
+
+        public void AddConnector(DiagramContentItem diagramContentItemToConnectTo)
```
Comment (samatrhea) [inline CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs:171]: docs

File: CDP4Grapher/GrapherModule.cs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        /// <summary>
+        /// Gets the <see cref="IThingDialogNavigationService"/> that handles navigation to dialogs
+        /// </summary>
+        internal IThingDialogNavigationService ThingDialogNavigationService { get; private set; }
+
+        /// <summary>
+        /// Gets the <see cref="IPluginSettingsService"/> used to read and write plugin setting files.
+        /// </summary>
+        internal IPluginSettingsService PluginSettingsService { get; private set; }
+
```
Comment (samatrhea) [inline CDP4Grapher/GrapherModule.cs:76]: lines

File: CDP4Grapher/Models/Node.cs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+//    The CDP4-IME Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//    GNU Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
```
Comment (samatrhea) [inline CDP4Grapher/Models/Node.cs:26]: lines

File: CDP4Grapher/ViewModels/GrapherViewModel.cs [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+                                objectChange.ChangedThing.RevisionNumber > this.RevisionNumber)
+                        .ObserveOn(RxApp.MainThreadScheduler)
+                        .Subscribe(x => this.UpdateProperties());
+                }
+                this.subscriptions.Add(listener);
+
+                this.GraphElements.Add(new GraphElementViewModel(nestedElement));
+            }
+
+            //this.NestedElements.AddRange(elements);
```
Comment (samatrhea) [inline CDP4Grapher/ViewModels/GrapherViewModel.cs:280]: please remove commented code

File: CDP4Grapher/ViewModels/GrapherRibbonViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        /// <param name="pluginSettingsService">
+        /// The <see cref="IPluginSettingsService"/> used to read and write plugin setting files.
+        /// </param>
+        /// <returns>An instance of <see cref="GrapherViewModel"/></returns>
+        public static GrapherViewModel InstantiatePanelViewModel(Option option, ISession session, IThingDialogNavigationService thingDialogNavigationService, IPanelNavigationService panelNavigationService, IDialogNavigationService dialogNavigationService, IPluginSettingsService pluginSettingsService)
+        {
+            return new GrapherViewModel(option, session, thingDialogNavigationService, panelNavigationService, dialogNavigationService, pluginSettingsService);
+        }
+    }
+}
```
Comment (lxatrhea) [inline CDP4Grapher/ViewModels/GrapherRibbonViewModel.cs:67]: NewLine please

File: CDP4Grapher/ViewModels/GrapherViewModel.cs [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+        /// </summary>
+        private void ClearNestedElementsAndDisposeSubscriptions()
+        {
+            this.subscriptions.ForEach(x => x.Dispose());
+            this.GraphElements.ForEach(x => x.NestedElementElementListener.Dispose());
+            this.subscriptions.Clear();
+            this.GraphElements.Clear();
+        }
+    }
+}
```
Comment (lxatrhea) [inline CDP4Grapher/ViewModels/GrapherViewModel.cs:321]: NewLine please

File: CDP4Grapher/Views/Grapher.xaml.cs [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        /// </remarks>
+        public Grapher(bool initializeComponent)
+        {
+            if (initializeComponent)
+            {
+                this.InitializeComponent();
+            }
+        }
+    }
+}
```
Comment (lxatrhea) [inline CDP4Grapher/Views/Grapher.xaml.cs:63]: NewLine please

File: CDP4Grapher/Views/GrapherDiagramControl.xaml.cs [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+            this.InitializeComponent();
+        }
+
+        /// <summary>
+        /// Override the Dev-Express context-menu
+        /// </summary>
+        /// <returns>The context-menu</returns>
+        protected override IEnumerable<IBarManagerControllerAction> CreateContextMenu() => ((IGrapherViewModel)this.DataContext).DiagramContextMenuViewModel.ContextMenu;
+    }
+}
```
Comment (lxatrhea) [inline CDP4Grapher/Views/GrapherDiagramControl.xaml.cs:71]: NewLine please

File: CDP4Grapher/Views/GrapherRibbon.xaml.cs [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        /// Initializes a new instance of the <see cref="GrapherRibbon"/> class
+        /// </summary>
+        [ImportingConstructor]
+        public GrapherRibbon()
+        {
+            this.InitializeComponent();
+            this.DataContext = new GrapherRibbonViewModel();
+        }
+    }
+}
```
Comment (lxatrhea) [inline CDP4Grapher/Views/GrapherRibbon.xaml.cs:53]: NewLine please

File: CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+                    throw new ArgumentOutOfRangeException(nameof(layout), layout, $"the combination between {layout} and {direction} is invalid");
+            }
+
+            this.CurrentLayout = (layout, direction);
+        }
+
+        /// <summary>
+        /// Raised when the item collection has changed
+        /// </summary>
+        /// <param name="sender"></param>
```
Comment (alexatrhea) [inline CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs:129]: empty comments

File: CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+        /// Applies the Circular Layout onto the diagram elements
+        /// </summary>
+        public void ApplyCircularLayout()
+        {
+            this.AssociatedObject.ApplyCircularLayout(this.AssociatedObject.Items);
+        }
+
+        /// <summary>
+        /// Applies the Tip over Layout onto the diagram elements
+        /// <param name="direction">the direction of the layout</param>
```
Comment (alexatrhea) [inline CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs:259]: param should be outside summary

File: CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+        /// Applies the Org Chart Layout onto the diagram elements
+        /// </summary>
+        public void ApplyOrgChartLayout()
+        {
+            this.AssociatedObject.ApplyOrgChartLayout(this.AssociatedObject.Items, SplitToConnectedComponentsMode.AllComponents);
+        }
+
+        /// <summary>
+        /// Applies the Mind map Layout onto the diagram elements
+        /// <param name="direction">the direction of the layout</param>
```
Comment (alexatrhea) [inline CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs:276]: param outside of summary

File: CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+        /// </summary>
+        public void ApplyMindMapLayout(OrientationKind direction)
+        {
+            this.AssociatedObject.ApplyMindMapTreeLayout(direction, this.AssociatedObject.Items, SplitToConnectedComponentsMode.AllComponents);
+        }
+
+        /// <summary>
+        /// Applies the Fugiyama Layout onto the diagram elements
+        /// <param name="direction">the direction of the layout</param>
```
Comment (alexatrhea) [inline CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs:285]: same as above

File: CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+        /// </summary>
+        public void ApplySugiyamaLayout(Direction direction)
+        {
+            this.AssociatedObject.ApplySugiyamaLayout(direction, this.AssociatedObject.Items);
+        }
+
+        /// <summary>
+        /// Applies the Tree Layout onto the diagram elements
+        /// <param name="direction">the direction of the layout</param>
```
Comment (alexatrhea) [inline CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs:294]: Same as above

File: CDP4Grapher/Behaviors/IGrapherOrgChartBehavior.cs [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        /// <summary>
+        /// Export the graph as the specified <see cref="DiagramExportFormat"/>
+        /// </summary>
+        /// <param name="format">the format to export the diagram to</param>
+        void ExportGraph(DiagramExportFormat format);
+
+        /// <summary>
+        /// Apply the desired layout specified
+        /// <param name="layout">the <see cref="LayoutEnumeration"/> layout to apply </param>
```
Comment (alexatrhea) [inline CDP4Grapher/Behaviors/IGrapherOrgChartBehavior.cs:53]: same as above, summary should close before the params for this file entirely

File: CDP4Grapher/Utilities/GrapherSaveFileDialog.cs [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+
+        /// <summary>
+        /// Gets or Sets the output file format
+        /// </summary>
+        public DiagramExportFormat Format { get; private set; }
+
+        /// <summary>
+        /// Instanciate a new <see cref="GrapherSaveFileDialog"/>
+        /// </summary>
+        /// <param name="format"></param>
```
Comment (alexatrhea) [inline CDP4Grapher/Utilities/GrapherSaveFileDialog.cs:58]: empty param

File: CDP4Grapher/ViewModels/DiagramControlContextMenuViewModel.cs [hunk 23]
```diff
@@ -0,0 +1,289 @@
...
+        public ReactiveCommand<object> ApplyMindMapLayoutLeftToRight { get; private set; }
+
+        /// <summary>
+        /// Gets the <see cref="ReactiveCommand"/> that apply the Org chart layout to the diagram elements
+        /// </summary>
+        public ReactiveCommand<object> ApplyOrganisationalChartLayout { get; private set; }
+
+        /// <summary>
+        /// Instiate a new <see cref="DiagramControlContextMenuViewModel"/>
+        /// <param name="behavior">the <see cref="IGrapherOrgChartBehavior"/> behavior</param>
```
Comment (alexatrhea) [inline CDP4Grapher/ViewModels/DiagramControlContextMenuViewModel.cs:156]: same as before

File: CDP4Grapher/ViewModels/DiagramControlContextMenuViewModel.cs [hunk 24]
```diff
@@ -0,0 +1,289 @@
...
+            this.ApplyTipOverLayoutRightToLeft.Subscribe(_ => this.Behavior.ApplySpecifiedLayout(LayoutEnumeration.TipOver, TipOverDirection.RightToLeft));
+
+            this.ApplyOrganisationalChartLayout = ReactiveCommand.Create();
+            this.ApplyOrganisationalChartLayout.Subscribe(_ => this.Behavior.ApplySpecifiedLayout(LayoutEnumeration.OrganisationalChart));
+
+            this.ApplyCircularLayout = ReactiveCommand.Create();
+            this.ApplyCircularLayout.Subscribe(_ => this.Behavior.ApplySpecifiedLayout(LayoutEnumeration.Circular));
+        }
+        
+        private void CreateContextMenu()
```
Comment (alexatrhea) [inline CDP4Grapher/ViewModels/DiagramControlContextMenuViewModel.cs:211]: missing docs

File: CDP4Grapher/ViewModels/GraphElementViewModel.cs [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+//    The CDP4-IME Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//    GNU Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
```
Comment (alexatrhea) [inline CDP4Grapher/ViewModels/GraphElementViewModel.cs:25]: unnecessary newline

File: CDP4Grapher/Utilities/SvgHelper.cs [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+//    The CDP4-IME Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//    GNU Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
```
Comment (alexatrhea) [inline CDP4Grapher/Utilities/SvgHelper.cs:25]: unnecessary new line

File: CDP4Grapher/Views/Grapher.xaml.cs [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+//    The CDP4-IME Community Edition is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//    GNU Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+namespace CDP4Grapher.Views
```
Comment (alexatrhea) [inline CDP4Grapher/Views/Grapher.xaml.cs:25]: newline please

**Comment by alexatrhea** [comment]
Converted to draft now to ensure we dont merge in before 6.0.0

**Comment by alexatrhea** [comment]
So please still do the Installer integration still into this branch

File: CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs [hunk 28]
```diff
@@ -0,0 +1,328 @@
...
+
+        /// <summary>
+        /// Holds the value whether the <see cref="DiagramControl"/> has loaded for the first time
+        /// </summary>
+        private bool hasLoaded;
+
+        /// <summary>
+        /// The on attached event handler
+        /// </summary>
+        protected override void OnAttached()
```
Comment (samatrhea) [inline CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs:64]: i would expect to see a OnDetaching as well to lean up the event handlers

File: CDP4Grapher/Utilities/GrapherSaveFileDialog.cs [hunk 29]
```diff
@@ -0,0 +1,84 @@
...
+    using System.IO;
+
+    using DevExpress.Diagram.Core;
+
+    using Microsoft.Win32;
+
+    /// <summary>
+    /// Represent a Save File dialog that will save as the outputs of the Grapher tool
+    /// </summary>
+    public class GrapherSaveFileDialog : IGrapherSaveFileDialog
```
Comment (samatrhea) [inline CDP4Grapher/Utilities/GrapherSaveFileDialog.cs:38]: We typically use the `IOpenSaveFileDialogService` to manage this. Please have a look at `LogInfoPanelViewModel` in the `CDP4LogInfo` project to get inspiration

File: CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs (hunk 28 above)
Comment (nathanatrhea) [reply CDP4Grapher/Behaviors/GrapherOrgChartBehavior.cs:64]: It's here on line 183

File: CDP4Grapher/Utilities/GrapherSaveFileDialog.cs (hunk 29 above)
Comment (nathanatrhea) [reply CDP4Grapher/Utilities/GrapherSaveFileDialog.cs]: thanks for the example, that helped

**Pull request merged**